src/hardware/mixer.cpp

```diff
@@ -1693,45 +1694,130 @@ AudioFrame MixerChannel::ApplyCrossfeed(const AudioFrame frame) const
+			                            *fade_ms <= max_duration_ms);
+			if (wait_is_valid && fade_is_valid) {
+				set_wait_and_fade(*wait_ms, *fade_ms);
+				wants_fadeout = true;
+				return;
+			}
+		}
+	}
+	// Otherwise inform the user and disable the fade
+	LOG_WARNING("%s: Invalid custom fadeout definition: '%s'. Must be "
+	            "specified in \"WAIT FADE\" format. Both values need "
+	            "to be between %d and %d (milliseconds). Using 'off'",
+	            channel.GetName().c_str(),
+	            prefs.data(),
+	            min_duration_ms,
+	            max_duration_ms);
+
+	wants_fadeout = false;
+}
+
+void MixerChannel::Sleeper::DecrementFadeLevel(const int64_t awake_for_ms)
+{
+	assert(awake_for_ms >= fadeout_or_sleep_after_ms);
+	const auto elapsed_fade_ms = static_cast<float>(
+	        awake_for_ms - fadeout_or_sleep_after_ms);
 
-	constexpr auto silence_threshold = 1.0f;
+	const auto decrement = fadeout_decrement_per_ms * elapsed_fade_ms;
 
-	had_noise = fabsf(frame.left) > silence_threshold ||
-	            fabsf(frame.right) > silence_threshold;
+	constexpr auto min_level = 0.0f;
+	constexpr auto max_level = 1.0f;
+	fadeout_level = std::clamp(max_level - decrement, min_level, max_level);
 }
 
-void MixerChannel::Sleeper::MaybeSleep()
+MixerChannel::Sleeper::Sleeper(MixerChannel& c, const uint16_t sleep_after_ms)
+        : channel(c),
+          fadeout_or_sleep_after_ms(sleep_after_ms)
 {
-	constexpr auto consider_sleeping_after_ms = 250;
+	// The constructed sleep period is programmatically controlled (so assert)
+	assert(fadeout_or_sleep_after_ms >= min_duration_ms);
+	assert(fadeout_or_sleep_after_ms <= max_duration_ms);
+}
 
-	// Not enough time has passed.. try again later
-	if (GetTicksSince(woken_at_ms) < consider_sleeping_after_ms) {
-		return;
+// Either fades the frame by the current fade-level, or records if the frame had
+// a magnitude great than 1. Cnce "had_noise" is tripped, it can only be reset
+// after waking up.
+void MixerChannel::Sleeper::MaybeFadeOrListen(AudioFrame& frame)
+{
+	if (wants_fadeout) {
+		// When fading, we actively drive down the channel level
+		frame.left *= fadeout_level;
+		frame.right *= fadeout_level;
+	} else if (!had_noise) {
+		// Otherwise, we simply passively listen for noise from the channel
+		constexpr auto silence_threshold = 1.0f;
+		had_noise = fabsf(frame.left) > silence_threshold ||
+		            fabsf(frame.right) > silence_threshold;
 	}
+}
 
-	// Stay awake if it's been noisy, otherwise we can sleep
-	if (had_noise) {
+void MixerChannel::Sleeper::MaybeSleep()
+{
+	const auto awake_for_ms = GetTicksSince(woken_at_ms);
+	// Not enough time has passed.. .. try to sleep later
+	if (awake_for_ms < fadeout_or_sleep_after_ms) {
+		return;
+	}
+	if (wants_fadeout) {
+		// The channel is still fading.. try to sleep later
+		if (fadeout_level > 0.0f) {
+			DecrementFadeLevel(awake_for_ms);
+			return;
+		}
+	} else if (had_noise) {
```

Comment:
Existing code but `had_noise` and the use of the term "noise" gave me a pause. I know this is the "sleeper" class but still, we're in the context of OPL and Tandy here which feature noise waveforms and dedicated noise channels, so I find the use of the word "noise" instead of "signal" confusing.

I recommend using a less confusing term such as "level" or "signal" or something.

Reply:
`had_signal` looks better; switched!